만국박람회 [STEP 3] 리지, Rowan

yijiye opened the pull request:
@yim2627 
안녕하세요 지성!! step3 완성하여 PR보냅니다 😀 
잘 부탁드립니다!

# 만국박람회 STEP 3 PR
# 고민했던 점 
## 첫 번째 화면 세로로 고정하기
특정 화면만 portrait orientation으로 고정하기 위해 여러가지 방법을 찾아보았습니다. 저희는 그 중에서 화면의 orientation에 대한 상태에 응답하는 것이 중요한 부분이라고 생각하여 AppDelegate를 활용한 방법을 선택하여 구현해 보았습니다.

```swift
class AppDelegate: UIResponder, UIApplicationDelegate {

    var shouldSupportAllOrientation = true

    func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
        shouldSupportAllOrientation == true ? .all : .portrait
    }
}
```

```swift
final class ExpoViewController: UIViewController {
   
    private let appDelegate = UIApplication.shared.delegate as? AppDelegate
}
```

<br/>

## TableViewCell Contents Accessibility 설정하기
테이블 뷰에 VoiceOver를 실행해보니 Cell을 통째로 인식해서 내부의 Label Title이 모두 합쳐져 cell.accessibilityLabel에 설정되어 있는 것을 확인했습니다.

cell 내부의 정보를 사용자에게 VoiceOver를 통해 들려주는 것이 바람직하다고 생각하여 해결 방법을 찾아보았습니다.

비공식 프로토콜 `UIAccessibilityContainer`를 채택하고 있는 클래스는 해당 클래스의 subview를 선택적으로 accessibiliyElement로 만들어줄 수 있는 메서드 / 프로퍼티를 제공하고 있습니다. 그 중에서 `var accessibilityElements: [Any]?`을 이용하였습니다.

1. cell 자체에 VoiceOver가 접근하지 못하도록 해준다.
2. cell.accessibilityElements에 label, imageView를 추가한다.
3. 각각의 subview들의 accessibilityProperty를 설정한다.

```swift
func setAccessibilityProperties(itemName: String, shortDescription: String) {
    self.itemNameLabel.accessibilityLabel = itemName
        
    self.shortDescriptionLabel.accessibilityLabel = "짧은 설명"
    self.shortDescriptionLabel.accessibilityValue = shortDescription
        
    self.itemImageView.isAccessibilityElement = true
    self.itemImageView.accessibilityLabel = itemName
    self.itemImageView.accessibilityTraits = .image
    self.itemImageView.accessibilityHint = "현재 셀을 선택하면 상세 화면으로 이동합니다."
        
    self.isAccessibilityElement = false
    self.accessibilityElements = [
        self.itemNameLabel!,
        self.shortDescriptionLabel!,
        self.itemImageView!
    ]
}
```

<br/>

# 조언이 필요한 점

## Description 화면 가로회전 시 height가 더 긴 이미지의 오토레이아웃 

- Description 화면에서 `imageView`와 `label`을 `stackView`에 넣어 오토레이아웃을 적용하였습니다. 이때 imageView에 들어오는 image의 크기에 따라 intrinsic size가 결정되어 어느곳을 기준으로 오토레이아웃을 잡을지 고민하였습니다. 
처음 선택한 방법은 아래 사진처럼 `label`의 `Top`을 `Contents view` 기준으로 150 수치만큼 떨어트려 그 공간만큼 `image`가 맞춰지도록 하였습니다. 그러나 화면을 가로로 회전할 때 `image` 크기가 화면에 맞게 늘어나지 않는 문제가 있었습니다.

<img src="https://i.imgur.com/pb8NkIk.png" width="400">

<br/>

- 그래서 그 문제를 해결하고자 두번째 방법(현재 적용방법)으로 `imageView`의 오토레이아웃을 `Frame Layout Guide`와 `width` 비율을 주고 `imageView`의 ratio를 1:1로 지정하였습니다. 이렇게 할 경우, 화면을 가로로 회전할 때 화면 크기에 맞춰 이미지 크기도 같이 증가하게 되었습니다. 그러나 이미지 크기별로 label간의 간격이 다 달라 이부분을 어떻게 해결할 수 있을지 지성의 조언이 필요합니다! 

<img src="https://i.imgur.com/gPcn92c.png" width="400">

<br/>

## Expo화면에서 numberOfLines에 따른 Dynamic type 오류

Dynamic type을 적용하고자 방문객, 개최지, 개최 기간과 그 설명을 다른 두개의 `label`로 나누어 글씨체를 다르게 적용하였습니다. 그리고 글씨크기가 커지면 자동줄바꿈 기능을 주고자 line 수를 0으로 지정하였는데 두개의 `label` 둘다 0으로 주었더니 오토레이아웃을 잡았어도 `label`간 간격이 벌어지는 문제가 발생하였습니다.
따라서 앞의 `label (방문객, 개최지, 개최기간)`은 글자 수가 많지 않아 글씨크기가 커져도 줄바꿈이 필요없어 line을 1로 하고 뒤에 길어지는 `label`의 line만 0으로 지정하였습니다.
오류는 해결하였지만, 왜 첫 번째 label의 line을 0으로 주었을 때, 여백이 생기는 이유는 잘 모르겠습니다. 어떤 부분을 고민해보면 좋을까요? 🤔

<img src="https://i.imgur.com/yOs4vmQ.png" width="300">

- 방문객 label (line: 0) / 방문객수 label (line: 0) 인 경우 공백 문제 발생.
- 개최지 label (line: 1) / 프랑스 파리 label (line: 0) 인 경우 문제 없음.

<br/>

# 해결하지 못한 점
## localization base 파일 오류
Accessibility 속성들을 정의하여 Accessibility Inspector를 통해 VoiceOver 기능의 시뮬레이션을 진행했습니다. 이때 활동학습 시간에 localization을 korean으로 설정하면 "button" 등의 기본 속성이 "버튼"으로 바뀌는 것을 시도해보고자 localization을 설정해주었는데요... 해당 변경사항을 원래대로 되돌리고 싶어 localization 속성을 삭제했더니 첫 번째 화면의 "한국의 출품작 보러가기" 버튼의 타이틀이 "Button"으로 고정되는 현상이 발생했습니다. 

저희는 의문이 들었습니다.. Xcode에서 String file을 삭제한 것 같은데도 계속해서 버튼의 타이틀이 Button 으로 고정되는 이유가 뭘까.. 추측하기로는 String file을 Xcode 내부적으로 갖고 있어서 따로 변경하지 않으면 해당 사항이 계속 저장된 상태로 남아있는 것 같습니다.

[Localization Tutorial](https://medium.com/lean-localization/ios-localization-tutorial-938231f9f881)이라는 글을 발견해서 알아보니 `Localizable.strings` 이라는 file을 통해 localization을 구현해 주는 것이라고 이해하게 되었습니다. 해당 파일은 기본적으로 생성되지 않으며 수정하고 싶을 때 개발자가 따로 Resource tab의 Strings File을 선택해 생성해야 한다고 합니다.

저희가 생각하고 있는게 맞을까요? 그리고 String file의 설정 기록을 남기지 않고 삭제하는 방법은 없을까요? 😢

일단 이번 문제는 git에서 reset, merge를 이용해 문제가 발생하기 이전 커밋 상태로 돌아가 해결했습니다.

<br/> 

## NavigationBar backbutton Accessibility 적용하는 방법

두 번째 화면에서 첫 번째 화면으로 돌아갈 때, navigationBar `<메인` 버튼을 눌러서 이동하게 되는데 이 버튼을 눌렀을 때 AccessibilityHint를 적용하려고 했습니다.
그래서 시도한 방법으로
- 두번째 화면의 setNavigationBar 메서드에 추가

- 메인이 곧 첫번째 화면의 navigationBarTitle 이므로 첫번째 화면의 setNavigationBar() 메서드에 추가

```swift
self.navigationItem.backBarButtonItem?.accessibilityHint
``` 

위의 두가지 방법을 시도하였으나 제대로 작동하지 않았습니다. 방법을 찾아보니 저희와 같은 고민을 한 사람은 많았지만 제대로 된 해결법을 찾지 못하여 지성에게 조언을 구하고 싶습니다..!

yijiye committed:
refactor: Expo 화면 스토리보드 오토레이아웃 적용

Tediousday93 committed:
feat: AppDelegate를 이용한 ExpoViewController 화면 세로 고정 구현

yijiye committed:
refactor: ItemEntryViewController 화면 오토레이아웃 수정

Tediousday93 committed:
refactor: DescriptionViewController 화면 오토레이아웃 수정

Tediousday93 committed:
refactor: 불필요한 코드 삭제 및 IBOutlet 네이밍 수정

Tediousday93 committed:
refactor: 사용하지 않는 메서드 삭제

yijiye committed:
refactor: 매직리터럴 열거형으로 처리

Tediousday93 committed:
rerfactor: tableviewCell contents 오토레이아웃 수정

가로모드에서 이미지 사이즈가 조절되지 않는 부분 개선

yijiye committed:
feat: Expo화면 VoiceOver 구현

Tediousday93 committed:
feat: CustomCell setAccessibilityProperties() 메서드 구현

yijiye committed:
feat: ItemEntry 화면 VoiceOver 구현

Tediousday93 committed:
feat: Description 화면 VoiceOver 구현

yijiye committed:
refactor: Description 화면 오토레이아웃 수정

Tediousday93 committed:
refactor: Accessibility 관련 메서드 기능 분리 및 컨벤션 일치하도록 코드 수정

setAccessibilityProperties() 의 수행 객체가 ViewController로 통일될 수 있도록 수정

yijiye committed:
refactor: DecodeManager method 수정

yijiye committed:
refactor: decodeData method 파라미터 이름 수정

Tediousday93 committed:
Merge remote-tracking branch 'refs/remotes/origin/step3'

Conflicts:
	Expo1900/Expo1900/ViewController/DecodeManager.swift
	Expo1900/Expo1900/ViewController/ExpoViewController.swift
	Expo1900/Expo1900/ViewController/ItemEntryViewController.swift

Tediousday93 committed:
docs: README.md 수정

yijiye committed:
docs: README.md 업데이트

yim2627 commented on Expo1900/Expo1900/App/AppDelegate.swift:
```diff
@@ -9,8 +9,12 @@ import UIKit
-
+    func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
+        shouldSupportAllOrientation == true ? .all : .portrait
+    }
```
앱 전역을 담당하는 타입이기때문에 특정 화면을 지정하기엔 적합하지 않다는 생각이드네요.
많은 고민을 하신 것은 좋네요!
`UINavigationControllerDelegate`에 대해서 알아보시면 될 것 같습니다.

Tediousday93 replied:
`UINavigationControllerDelegate` 검색해도 나오지 않던 정보네요 👍 가르쳐주셔서 감사합니다 !!

> 앱 전역을 담당하는 타입이기때문에 특정 화면을 지정하기엔 적합하지 않다는 생각이드네요.

저희도 지성의 생각에 동의합니다!
그래서, 추가로 해당 Delegate에 대한 학습 진행해보고 코드에 적용시켜 봤습니다.

```swift
final class ExpoViewController: UIViewController {
    private func setNavigationControllerDelegate() {
        self.navigationController?.delegate = self
    }
}

extension ExpoViewController: UINavigationControllerDelegate {
    func navigationControllerSupportedInterfaceOrientations(
        _ navigationController: UINavigationController
    ) -> UIInterfaceOrientationMask {
        return .portrait
    }
}
```

yim2627 commented on Expo1900/Expo1900/ViewController/DescriptionViewController.swift:
```diff
@@ -12,13 +12,14 @@ final class DescriptionViewController: UIViewController {
+    @IBOutlet private weak var itemImageView: UIImageView!
     
     override func viewDidLoad() {
         super.viewDidLoad()
+        setNavigationBar()
         setImageView()
         setLabel()
-        setNavigationBar()
+        setAccessibilityProperties()
     }
```
👍

yim2627 commented on Expo1900/Expo1900/ViewController/ExpoViewController.swift:
```diff
@@ -81,16 +92,6 @@ final class ExpoViewController: UIViewController {
         return decimalVisitors
     }
     
-    private func convertTextSize(of label: UILabel, range: NSRange) {
-        guard let target = label.text else { return }
-        
-        let fontSize = UIFont.systemFont(ofSize: 20)
-        let attributedString = NSMutableAttributedString(string: target)
-        
-        attributedString.addAttribute(.font, value: fontSize, range: range)
-        label.attributedText = attributedString
-    }
```
왜 제거하신건가요?

yijiye replied:
첫 번째 화면의 label을 두개로 나누면서 하나의 label 안에서 특정 글씨 크기만 조정할 필요가 없어졌습니다. 그래서 위의 메서드를 삭제했었습니다.
그런데, 질문남긴 내용중 label의 line에 따라 공백이 생기는 원인이 오토레이아웃 문제라 저희도 생각하여 (여러가지 다 시도했는데 label 이 두개일 경우는 결국 해답을 찾지못했습니다 🥲) label을 다시 하나로 변경하였습니다. 
결론적으로 위의 메서드를 다시 사용하였습니다...!! 
[2b43396](https://github.com/yagom-academy/ios-exposition-universelle/pull/257/commits/2b433960989310c11d24c9518c4ed9c3823972dd)

yim2627 commented on Expo1900/Expo1900/ViewController/ItemEntryViewController.swift:
```diff
@@ -55,17 +55,14 @@ extension ItemEntryViewController: UITableViewDataSource {
+        setAccessibilityProperties(of: cell, itemName: itemName, shortDescription: shortDescription)
         cell.accessoryType = .disclosureIndicator
     }
 }
```
현재 ViewController에 Initializer로 주입을 해주시고 계신데, Cell은 왜 ViewController까지 갖고와서 해주시는건가요? 현재 해주신 것처럼 메서드를 통해 아이템을 주입시켜주고 해당 뷰(Cell)에서 셋팅하는 것이 더 명확하지않을까요?
커스텀 셀쪽의 코드는 뷰만 있고 아무것도 없어서 질문드립니다.

Tediousday93 replied:
UITableViewCell에서 `setContents`를 정의하는 것이 명확하다고 생각하여 `setContents` 메서드의 위치를 cell 타입 내부로 수정하였습니다!

Item을 주입해주는 메서드를 추가하여 dequeueReusableCell 이후 데이터를 주입받도록 하였습니다.

setContents를 옮기면서 Accessibility 관련 메서드도 Cell로 옮겨주는게 맞는 것 같아 같이 수정했습니다!

yim2627 commented on Expo1900/Expo1900/ViewController/ItemEntryViewController.swift:
```diff
@@ -87,3 +84,38 @@ extension ItemEntryViewController: UITableViewDelegate {
+            cell.itemNameLabel!,
+            cell.shortDescriptionLabel!,
+            cell.itemImageView!
+        ]
```
강제 언래핑을 해주신 이유가 있을까요? 강제 언래핑의 위험성을 알아보셔도 좋을 것 같습니다.
그리고 셀쪽에서 담당하는 것이 더 명확하다는 생각이 드는데 어떻게 생각하시나요?

yijiye replied:
넵 이부분은 테스트로 달아놓고 그냥 PR을 보내버렸네요 😅
옵셔널 바인딩으로 해주려다가 content들을 @IBOutlet collection으로 묶어서 forEach를 사용하여 accessibilityElements에 넣어주는 방향으로 풀어보았습니다!

yim2627 replied:
개인적으로 Cell 내부로 해당 코드를 옮겨도 될 것 같다는 생각이 드네요!

yim2627 commented on Expo1900/Expo1900/ViewController/ExpoViewController.swift:
```diff
@@ -104,5 +105,37 @@ final class ExpoViewController: UIViewController {
+        self.flagImageViews[0].accessibilityLabel = "버튼 오른쪽 태극기"
+        self.flagImageViews[1].accessibilityLabel = "버튼 왼쪽 태극기"
+    }
+}
```
각자 취향이지만, 익스텐션 자체에 Private을 붙이는 방법도 있습니다. 그럼 익스텐션 내부의 메서드들도 전부 Private이 됩니다.

yim2627 commented on Expo1900/Expo1900/App/AppDelegate.swift:
```diff
@@ -9,8 +9,12 @@ import UIKit
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
```
final을 붙여도 될 것 같습니다.

yim2627 reviewed (commented):
@yim2627 
안녕하세요 지성!! step3 완성하여 PR보냅니다 😀 
잘 부탁드립니다!

# 만국박람회 STEP 3 PR
# 고민했던 점 
## 첫 번째 화면 세로로 고정하기
특정 화면만 portrait orientation으로 고정하기 위해 여러가지 방법을 찾아보았습니다. 저희는 그 중에서 화면의 orientation에 대한 상태에 응답하는 것이 중요한 부분이라고 생각하여 AppDelegate를 활용한 방법을 선택하여 구현해 보았습니다.

```swift
class AppDelegate: UIResponder, UIApplicationDelegate {

    var shouldSupportAllOrientation = true

    func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
        shouldSupportAllOrientation == true ? .all : .portrait
    }
}
```

```swift
final class ExpoViewController: UIViewController {
   
    private let appDelegate = UIApplication.shared.delegate as? AppDelegate
}
```

> 특정 화면만 건드는 거라면 건드는 로직을 특정 화면에 종속시키는 것이 좋을 것 같습니다.
<br/>

## TableViewCell Contents Accessibility 설정하기
테이블 뷰에 VoiceOver를 실행해보니 Cell을 통째로 인식해서 내부의 Label Title이 모두 합쳐져 cell.accessibilityLabel에 설정되어 있는 것을 확인했습니다.

cell 내부의 정보를 사용자에게 VoiceOver를 통해 들려주는 것이 바람직하다고 생각하여 해결 방법을 찾아보았습니다.

비공식 프로토콜 `UIAccessibilityContainer`를 채택하고 있는 클래스는 해당 클래스의 subview를 선택적으로 accessibiliyElement로 만들어줄 수 있는 메서드 / 프로퍼티를 제공하고 있습니다. 그 중에서 `var accessibilityElements: [Any]?`을 이용하였습니다.

1. cell 자체에 VoiceOver가 접근하지 못하도록 해준다.
2. cell.accessibilityElements에 label, imageView를 추가한다.
3. 각각의 subview들의 accessibilityProperty를 설정한다.

```swift
func setAccessibilityProperties(itemName: String, shortDescription: String) {
    self.itemNameLabel.accessibilityLabel = itemName
        
    self.shortDescriptionLabel.accessibilityLabel = "짧은 설명"
    self.shortDescriptionLabel.accessibilityValue = shortDescription
        
    self.itemImageView.isAccessibilityElement = true
    self.itemImageView.accessibilityLabel = itemName
    self.itemImageView.accessibilityTraits = .image
    self.itemImageView.accessibilityHint = "현재 셀을 선택하면 상세 화면으로 이동합니다."
        
    self.isAccessibilityElement = false
    self.accessibilityElements = [
        self.itemNameLabel!,
        self.shortDescriptionLabel!,
        self.itemImageView!
    ]
}
```

> 깊이까지 지원하려 하시는 점은 너무나도 좋지만, 강제 언래핑은 잘못된 방법 같습니다. 다른 방법은 무엇이 있을지 고민하시는 것이 좋을 것 같아요.

<br/>

# 조언이 필요한 점

## Description 화면 가로회전 시 height가 더 긴 이미지의 오토레이아웃 

- Description 화면에서 `imageView`와 `label`을 `stackView`에 넣어 오토레이아웃을 적용하였습니다. 이때 imageView에 들어오는 image의 크기에 따라 intrinsic size가 결정되어 어느곳을 기준으로 오토레이아웃을 잡을지 고민하였습니다. 
처음 선택한 방법은 아래 사진처럼 `label`의 `Top`을 `Contents view` 기준으로 150 수치만큼 떨어트려 그 공간만큼 `image`가 맞춰지도록 하였습니다. 그러나 화면을 가로로 회전할 때 `image` 크기가 화면에 맞게 늘어나지 않는 문제가 있었습니다.

<img src="https://i.imgur.com/pb8NkIk.png" width="400">

<br/>

- 그래서 그 문제를 해결하고자 두번째 방법(현재 적용방법)으로 `imageView`의 오토레이아웃을 `Frame Layout Guide`와 `width` 비율을 주고 `imageView`의 ratio를 1:1로 지정하였습니다. 이렇게 할 경우, 화면을 가로로 회전할 때 화면 크기에 맞춰 이미지 크기도 같이 증가하게 되었습니다. 그러나 이미지 크기별로 label간의 간격이 다 달라 이부분을 어떻게 해결할 수 있을지 지성의 조언이 필요합니다! 

<img src="https://i.imgur.com/gPcn92c.png" width="400">

> 우선 StackView 내부에서 오토레이아웃을 잡아주신 이유가 있을까요? 스택뷰의 distribution을 따라가기때문에 설정해주신 레이아웃이 무시되거나, 스택뷰 내부 레이아웃이 깨질 가능성이 있습니다.
제 생각엔 StackView에 이미지와 레이블을 넣어주실거면 이미지가 자신의 사이즈만큼 커지면서 레이블을 눌러 깨지지 않도록 우선 순위를 지정하면 좋을 것 같습니다.
<br/>

## Expo화면에서 numberOfLines에 따른 Dynamic type 오류

Dynamic type을 적용하고자 방문객, 개최지, 개최 기간과 그 설명을 다른 두개의 `label`로 나누어 글씨체를 다르게 적용하였습니다. 그리고 글씨크기가 커지면 자동줄바꿈 기능을 주고자 line 수를 0으로 지정하였는데 두개의 `label` 둘다 0으로 주었더니 오토레이아웃을 잡았어도 `label`간 간격이 벌어지는 문제가 발생하였습니다.
따라서 앞의 `label (방문객, 개최지, 개최기간)`은 글자 수가 많지 않아 글씨크기가 커져도 줄바꿈이 필요없어 line을 1로 하고 뒤에 길어지는 `label`의 line만 0으로 지정하였습니다.
오류는 해결하였지만, 왜 첫 번째 label의 line을 0으로 주었을 때, 여백이 생기는 이유는 잘 모르겠습니다. 어떤 부분을 고민해보면 좋을까요? 🤔

<img src="https://i.imgur.com/yOs4vmQ.png" width="300">

- 방문객 label (line: 0) / 방문객수 label (line: 0) 인 경우 공백 문제 발생.
- 개최지 label (line: 1) / 프랑스 파리 label (line: 0) 인 경우 문제 없음.

>  오토레이아웃이 문제일 가능성이 큽니다. 두 레이블을 스택뷰에 넣어주셨는데, 두 레이블의 우선순위를 매겨 해결할 수 있을 것 같습니다.
<br/>

# 해결하지 못한 점
## localization base 파일 오류
Accessibility 속성들을 정의하여 Accessibility Inspector를 통해 VoiceOver 기능의 시뮬레이션을 진행했습니다. 이때 활동학습 시간에 localization을 korean으로 설정하면 "button" 등의 기본 속성이 "버튼"으로 바뀌는 것을 시도해보고자 localization을 설정해주었는데요... 해당 변경사항을 원래대로 되돌리고 싶어 localization 속성을 삭제했더니 첫 번째 화면의 "한국의 출품작 보러가기" 버튼의 타이틀이 "Button"으로 고정되는 현상이 발생했습니다. 

저희는 의문이 들었습니다.. Xcode에서 String file을 삭제한 것 같은데도 계속해서 버튼의 타이틀이 Button 으로 고정되는 이유가 뭘까.. 추측하기로는 String file을 Xcode 내부적으로 갖고 있어서 따로 변경하지 않으면 해당 사항이 계속 저장된 상태로 남아있는 것 같습니다.

[Localization Tutorial](https://medium.com/lean-localization/ios-localization-tutorial-938231f9f881)이라는 글을 발견해서 알아보니 `Localizable.strings` 이라는 file을 통해 localization을 구현해 주는 것이라고 이해하게 되었습니다. 해당 파일은 기본적으로 생성되지 않으며 수정하고 싶을 때 개발자가 따로 Resource tab의 Strings File을 선택해 생성해야 한다고 합니다.

저희가 생각하고 있는게 맞을까요? 그리고 String file의 설정 기록을 남기지 않고 삭제하는 방법은 없을까요? 😢

일단 이번 문제는 git에서 reset, merge를 이용해 문제가 발생하기 이전 커밋 상태로 돌아가 해결했습니다.

> 우선 커스텀하게 로컬라이징을 진행하시려면, 튜토리얼에서 보신 것이 맞습니다. String file의 설정 기록을 남기지 않고 삭제하는 방법..은 잘 이해하지 못하겠네요. 변경된 것으로 기록이 남고 터미널이나 GUI에 보일테니 해당 데이터를 삭제해주면 되는 것 아닌가요?
<br/> 

## NavigationBar backbutton Accessibility 적용하는 방법

두 번째 화면에서 첫 번째 화면으로 돌아갈 때, navigationBar `<메인` 버튼을 눌러서 이동하게 되는데 이 버튼을 눌렀을 때 AccessibilityHint를 적용하려고 했습니다.
그래서 시도한 방법으로
- 두번째 화면의 setNavigationBar 메서드에 추가

- 메인이 곧 첫번째 화면의 navigationBarTitle 이므로 첫번째 화면의 setNavigationBar() 메서드에 추가

```swift
self.navigationItem.backBarButtonItem?.accessibilityHint
``` 

위의 두가지 방법을 시도하였으나 제대로 작동하지 않았습니다. 방법을 찾아보니 저희와 같은 고민을 한 사람은 많았지만 제대로 된 해결법을 찾지 못하여 지성에게 조언을 구하고 싶습니다..!

> `navigationController.navigationBar.backItem?.backBarButtonItem.accessibilityHint`으로 해결이 될지.. 모르겠네요 Back 이라고 알려줄텐데, 이정도면 사용자에게도 충분하지 않을까? 라는 생각도 들고요.. 어떻게 생각하시나요?

Tediousday93 committed:
refactor: Expo 화면 portrait orientation 고정 UINavigationControllerDelegate 활용하는 방법으로 변경

AppDelegate에 구현한 내용 삭제 ExpoViewController extension에 UINavigationControllerDelegate 채택

Tediousday93 committed:
Merge remote-tracking branch 'refs/remotes/origin/step3'

yijiye committed:
refactor: Expo 화면 Label 한개로 수정, 오토레이아웃 수정, 특정 Label 크기 조정

Tediousday93 committed:
refactor: setContents 메서드 위치 TabelViewCell 내부로 변경

yijiye committed:
refactor: setContents, setAccessibilityProperties method 위치 CustomTableViewCell로 변경

Tediousday93 committed:
refactor: 컨벤션 수정,  접근제어자 추가

yijiye commented:
> 우선 StackView 내부에서 오토레이아웃을 잡아주신 이유가 있을까요? 스택뷰의 distribution을 따라가기때문에 설정해주신 레이아웃이 무시되거나, 스택뷰 내부 레이아웃이 깨질 가능성이 있습니다.
제 생각엔 StackView에 이미지와 레이블을 넣어주실거면 이미지가 자신의 사이즈만큼 커지면서 레이블을 눌러 깨지지 않도록 우선 순위를 지정하면 좋을 것 같습니다.

지성이 알려주신 방법대로 적용을 해보았는데, imageView 자체에 제약을 걸지 않으면 사진의 크기 자체가 큰 image가 할당될때 height가 매우 길어지는 문제가 발생했습니다. 
<img width="324" alt="스크린샷 2023-03-04 오전 12 20 36" src="https://user-images.githubusercontent.com/114971172/222759760-1b6d3806-c4db-4f42-90ad-430ea9739065.png">

PR 작성전에도 여러가지 시도해본 결과 stackView의 제약만으로는 문제를 해결하기 어려워 imageView에 제약을 걸었습니다. 지성이 말씀하신 내용을 저희가 잘 이해하지 못한걸까요..? 어떻게 해결할 수 있을까요? 🥲

yim2627 reviewed (commented):
큰사이즈의 이미지도 이미지 사이즈를 통일화 시켜주고 싶으시다면, 큰 사이즈의 이미지를 기준으로 잡고 이미지의 ratio를 맞춘뒤 해당 비율을 넘어가지 않도록 lessThenEqul(>=)로 이미지 사이즈를 잡아주면 되겠다는 생각이 드네요. 보통 스택뷰의 distribution이 fill이면 내부에서 레이아웃을 잡지 않지만, 이런 케이스의 경우 이미지의 height을 한정시켜주는 것이 좋다는 생각이 드네요! 이견이 있으시다면 언제든지 DM이나,  답변 부탁드립니다.

yijiye committed:
refactor: Description 화면 imageView 오토레이아웃 수정

yijiye commented:
이미지들의 가로 세로 비율이 다달라서 기준을 잡기가 어려웠습니다. 그래서 imageView와 Frame layout guide의 proportional width( lessThenEqul(>=) 0.7), height( lessThenEqul(>=) 0.5) 제약을 걸어주는 방법으로 화면에 보여지는 imageView의 최대크기를 정해주었습니다!
